pkg/planner/core/indexmerge_path.go

```diff
@@ -596,7 +930,191 @@ func (ds *DataSource) generateIndexMergeOnDNF4MVIndex(normalPathCnt int, filters
+				IndexRangeScan(a, [2,2])              --- COP
+				IndexRangeScan(a, [3,3])              --- COP
+			IndexRangeScan(non-mv-index-if-any)(?)    --- COP
+			TableRowIdScan(t)                         --- COP
+*/
+func (ds *DataSource) generateIndexMerge4ComposedIndex(normalPathCnt int, indexMergeConds []expression.Expression) error {
+	isPossibleIdxMerge := len(indexMergeConds) > 0 && // have corresponding access conditions, and
+		len(ds.possibleAccessPaths) > 1 // have multiple index paths
+	if !isPossibleIdxMerge {
+		return nil
+	}
+
+	if len(indexMergeConds) == 1 {
+		// DNF path.
+		sf, ok := indexMergeConds[0].(*expression.ScalarFunction)
+		if !ok || sf.FuncName.L != ast.LogicOr {
+			// targeting: cond1 or cond2 or cond3
+			return nil
+		}
+		dnfFilters := expression.FlattenDNFConditions(sf)
+		mvIndexPartialPaths, usedAccessMap, needSelection4MVIndex, err := ds.generateIndexMergePartialPath4Or(normalPathCnt, dnfFilters)
+		if err != nil {
+			return err
+		}
+		if len(mvIndexPartialPaths) == 0 {
+			// no mv index partial paths to be composed of.
+			return nil
+		}
+		normalIndexPartialPaths, usedAccessMap, needSelection4NormalIndex, err := ds.generateNormalIndexPartialPath4Or(dnfFilters, usedAccessMap, normalPathCnt)
+		if err != nil {
+			return err
+		}
+		// since multi normal index merge path is handled before, here focus on multi mv index merge, or mv and normal mixed index merge
+		composed := (len(mvIndexPartialPaths) > 1) || (len(mvIndexPartialPaths) == 1 && len(normalIndexPartialPaths) >= 1)
+		if !composed {
+			return nil
+		}
+		// if any cnf item is not used as index partial path, index merge is not valid anymore.
+		for _, one := range usedAccessMap {
+			if !one {
+				return nil
+			}
+		}
+		// todo: make this code as a portal of all index merge path.
+		// if we derive:
+		// 1: some mv index partial path, no normal index path, it means multi mv index merge.
+		// 2: some mv index partial path, some normal index path, it means hybrid index merge.
+		// 3: no mv index partial path, several normal index path, it means multi normal index merge.
+		var combinedPartialPaths []*util.AccessPath
+		if len(normalIndexPartialPaths) != 0 {
+			combinedPartialPaths = normalIndexPartialPaths
+		}
+		if len(mvIndexPartialPaths) != 0 {
+			combinedPartialPaths = append(combinedPartialPaths, mvIndexPartialPaths...)
+		}
+		if len(combinedPartialPaths) == 0 {
+			return nil
+		}
+		// here we directly use the all index merge conditions as the table filers for simplicity.
+		// todo:  make estimation more correct rather than pruning other index merge path.
+		var indexMergeTableFilters []expression.Expression
+		if needSelection4MVIndex || needSelection4NormalIndex {
+			indexMergeTableFilters = indexMergeConds
+		}
+		mvp := ds.buildPartialPathUp4MVIndex(combinedPartialPaths, false, indexMergeTableFilters)
+		ds.possibleAccessPaths = append(ds.possibleAccessPaths, mvp)
+		return nil
+	}
+	// CNF path.
```

Comment:
What if we just extend the previous `generateIndexMergeAndPaths()` to treat the intersection type index merge path as a candidate partial path?
I think the new logic has exactly the same rationale as the previous one, but we are re-implementing many things.

Reply:
![image](https://github.com/pingcap/tidb/assets/15825830/06fb20dc-0efb-411d-a377-222cc33d896f)
From the pic, for normal index path in CNF case, I have tried to reuse this function, but for MV index conds, there some limitations, and that's why I specially split it out.

the way you say is the first version I have been tried, take this as a simple example
```
(1 member of a) and (2 member of a)
```
the old index merge case will only gen ONE candidate path for ONE MV index path for this, picking the first encountered/satisfied conds as it access conds, so for this case, there is exactly one index candidate partial path under extending generateIndexMergeAndPaths function as you said.

Reply:
![image](https://github.com/pingcap/tidb/assets/15825830/2e25f83c-52fb-4d4b-9eb0-2dbfc4ab6714)
there is a additional todo already added

Reply:
```
(1 member of a) and (2 member of a)
```
For this case, I think that means we should enhance the old mv index implementation.
Besides, I think this leads to another question: Are we relying on the new logic in this PR to handle this case?
If yes, what's the scope of this PR? I thought this PR is for generating index merge paths that contain different indexes (and including mv index).

Reply:
The answer is yes, it targeting for composition of multi different or same mv/normal index partial paths. and `(1 member of a) and (2 member of a)` is not mentioned in the user case, but it's done by the way. 😂